cupy/cuda/memory.pyx

```diff
@@ -1603,8 +1602,10 @@ cdef class MemoryPool:
     def __init__(self, allocator=None):
         if allocator is None:
             allocator = _malloc
-        self._pools = collections.defaultdict(
-            lambda: SingleDeviceMemoryPool(allocator))
+
+        dev_counts = runtime.getDeviceCount()
+        self._pools = tuple(
+            SingleDeviceMemoryPool(allocator) for i in range(dev_counts))
```

Comment:
This will probably fixed in Python, but doesn't hurt even now. It is the one true bug-fix here (except on future 3.14.x versions presumably).

It seems my very long odyssey was caused by `defaultdict`. The mechanism is that two threads can inintialize a defaultdict at the same time.
In that case, one thread will get a `SingleDeviceMemoryPool` that is immediately discarded again. Howver, `SingleDEviceMemoryPool` owns the actual allocations currently which causes havoc later.
(And at least often not all that obvious havoc...)